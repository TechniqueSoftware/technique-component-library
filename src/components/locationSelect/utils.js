export const urlClubIdRegex = id => new RegExp(`^.*\/${id}(\/)?(#)?(((\/)?(#){1}){1}([0-9a-zA-Z])+?)*?$`, 'gm') // eslint-disable-line no-useless-escape, max-len
const urlLocationIdRegex = id => new RegExp(`^.*\/${id}\/`, 'gm') // eslint-disable-line no-useless-escape

export const getLocationIdFromUrl = (path, clubId) => {
  try {
    const [urlBeginning] = path.match(urlLocationIdRegex(clubId))
    const locationId = path.substring(urlBeginning.length).replace(/[^0-9]/g, '')
    return locationId ? parseInt(locationId, 10) : null
  } catch (err) {
    return null
  }
}
