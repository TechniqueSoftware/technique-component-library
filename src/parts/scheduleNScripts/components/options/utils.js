export const followUpTimeToObject = followUpTime => {
  const getHour = h => {
    if (h === '00') return 12
    const parsedHour = parseInt(h, 10)
    return parsedHour > 12 ? parsedHour - 12 : parsedHour
  }
  const time = followUpTime.split(':')
  const [oldHour, oldMinute] = time
  const hour = getHour(oldHour)
  const minute = parseInt(oldMinute, 10)
  const meridiem = oldHour >= 12 ? 'PM' : 'AM'
  return { hour, minute, meridiem }
}

export const objectToFollowUpTime = (hour, minute, meridiem) => {
  const parseHour = h => meridiem === 'PM' ? parseInt(h, 10) + 12 : h
  return `${parseHour(hour === 12 ? '00' : hour)}:${minute === 0 ? '00' : minute}:00`
}
