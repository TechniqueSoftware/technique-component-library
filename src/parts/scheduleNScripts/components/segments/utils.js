import { FOLLOWUP_STATUS } from './constants'

export const parseTitle = followUpStatus => {
  switch (followUpStatus) {
    case FOLLOWUP_STATUS.WEB_LEAD:
      return 'Web Lead'
    case FOLLOWUP_STATUS.APPT_BOOKED:
      return 'Appt Booked'
    case FOLLOWUP_STATUS.MEMBER:
      return 'Member'
  }
}

export const parseLocations = segmentLocations => segmentLocations.map(({ locationName, locationId }) => ({
  label: `${locationName}`,
  value: locationId
}))

export const checkPermission = userPermission => userPermission.some(
  element => element.permissionName === 'CopySegmentsAction' && element.permissionEffect === 'ALLOW')
