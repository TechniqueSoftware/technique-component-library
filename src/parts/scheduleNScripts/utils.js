import { PERMISSION_KEYS, STATUSES_WITH_SEGMENTS } from './constants'

export const checkIfSegmentStatus = (type, id) => {
  const hasSegmentation = !!STATUSES_WITH_SEGMENTS.find(s => s.type === parseInt(type, 10) && s.id === parseInt(id, 10))
  return hasSegmentation
}

const setSchedulesDayNum = schedules => schedules.reduce(
  (acc, curr) => {
    const dayNum = curr.daysTillNext + acc.dayNum
    return { schedules: acc.schedules.concat({ ...curr, dayNum }), dayNum }
  },
  { dayNum: 0, schedules: [] }
).schedules

export const getStatusSegments = (type, id, segments) => {
  if (!checkIfSegmentStatus(type, id)) return []
  const currentStatus = segments.find(
    ({ followUpType, followUpStatus }) => followUpType === parseInt(type, 10) && followUpStatus === parseInt(id, 10)
  )
  if (!currentStatus) return []
  const { followUpSegments = [], followUpSchedules = [], followUpSegmentScripts = [] } = currentStatus
  return followUpSegments.map(segment => {
    const schedulesWithScripts = followUpSchedules.reduce((acc, schedule) => {
      if (segment.segmentId !== schedule.segmentId) return acc
      return acc.concat({
        ...schedule,
        scripts: followUpSegmentScripts.filter(
          script => script.followUpSeqNum === schedule.followUpSeqNum && script.segmentId === schedule.segmentId
        )
      })
    }, [])
    return { ...segment, followUpSchedules: setSchedulesDayNum(schedulesWithScripts) }
  })
}

export const sortSegments = segments => [...segments].sort((a, b) => a.name.localeCompare(b.name))

// Formatting club-os data into something usable
export const parseSnSData = (data, followUpSchedules, followUpScripts, segmentsData, isClub) => {
  const schedulesPermissions = followUpSchedules.permissions
  const scriptsPermissions = followUpScripts.permissions
  const schedules = isClub ? followUpSchedules : followUpSchedules.schedules
  const scripts = isClub ? followUpScripts : followUpScripts.scripts

  const group = (acc, curr) => {
    const { followUpStatus, followUpType } = curr
    const key = `${followUpStatus}-${followUpType}`
    if (acc[key]) acc[key].push(curr)
    else acc[key] = [curr]
    return acc
  }

  const sch = schedules.reduce(group, {})
  const scr = scripts.reduce(group, {})

  Object.keys(sch).forEach(key => { sch[key] = setSchedulesDayNum(sch[key]) })

  Object.keys(scr).forEach(key => {
    const scripts = scr[key]
    scripts.forEach(script => {
      const schedule = sch[key]?.find(el => el.followUpSeqNum === script.followUpSeqNum)
      if (!schedule) return
      if (schedule.scripts) schedule.scripts.push(script)
      else schedule.scripts = [script]
    })
  })

  const followups = data.map(el => {
    const { type, statuses, id } = el
    const canEditSchedules = isClub ? undefined : schedulesPermissions[PERMISSION_KEYS.SCHEDULES[type]]
    const canEditScripts = isClub ? undefined : scriptsPermissions[PERMISSION_KEYS.SCRIPTS[type]]
    const newStatuses = statuses.map(st => {
      const key = `${st.id}-${type}`
      const unsortedSegments = getStatusSegments(type, st.id, segmentsData)
      const segments = sortSegments(unsortedSegments)
      return { ...st, followUpTab: id, schedules: sch[key] || [], segments }
    })
    return { ...el, statuses: newStatuses, canEditSchedules, canEditScripts }
  })

  return {
    followups,
    canEditSchedules: followups[0].canEditSchedules,
    canEditScripts: followups[0].canEditScripts
  }
}

export const getSchedulesByDay = (schedules, dayNum, order) => {
  const { earlierSch, sameDaySch, laterSch } = schedules.reduce(
    (acc, sch) => {
      if (sch.dayNum < dayNum) acc.earlierSch = acc.earlierSch.concat(sch)
      if (sch.dayNum === dayNum) acc.sameDaySch = acc.sameDaySch.concat(sch)
      if (sch.dayNum > dayNum) acc.laterSch = acc.laterSch.concat(sch)
      return acc
    },
    { earlierSch: [], sameDaySch: [], laterSch: [] }
  )
  return {
    earlierSch: order === 'first' ? earlierSch : earlierSch.concat(sameDaySch),
    laterSch: order === 'first' ? sameDaySch.concat(laterSch) : laterSch
  }
}

export const getCurrentSchedules = (segmentId, status) => {
  if (!segmentId) return status.schedules
  return status.segments.find(segment => segment.segmentId === segmentId).followUpSchedules
}

export const getStatus = (followups, type, status) => {
  const fs = followups.filter(f => f.type === type)
  if (!fs.length) return null

  let st = null
  for (let i = 0; i < fs.length; i++) {
    st = fs[i].statuses.find(s => s.id === `${status}`)
    if (st) break
  }
  return st
}

export const setFollowUpStatusValue = (followUps, status, value) => {
  const newFollowups = followUps.map(followUp => {
    if (followUp.id !== status.followUpTab) return followUp
    const newStatuses = followUp.statuses.map(st => {
      if (st.id !== status.id) return st
      return { ...st, ...value }
    })
    return { ...followUp, statuses: newStatuses }
  })
  return newFollowups
}

export const setFollowUpScripts = (followUps, { followUpType, followUpStatus }, scriptsData, segmentId) => {
  const status = getStatus(followUps, followUpType, followUpStatus)
  const currentSchedules = getCurrentSchedules(segmentId, status)
  const schedules = currentSchedules.map(schedule => {
    if (schedule.followUpSeqNum !== scriptsData.followUpSeqNum) return schedule
    return scriptsData
  })
  const statusValueToUpdate = !segmentId
    ? { schedules }
    : {
      segments: status.segments.map(
        segment => (segment.segmentId !== segmentId ? segment : { ...segment, followUpSchedules: schedules }) // eslint-disable-line max-len
      )
    }
  return setFollowUpStatusValue(followUps, status, statusValueToUpdate)
}

export const updateTouchpointInFollowups = (followUps, tp, data) => {
  const { followUpType, followUpStatus, followUpSeqNum } = tp
  const status = getStatus(followUps, followUpType, followUpStatus)
  const newSchedules = status.schedules.map(sch => {
    if (sch.followUpSeqNum !== followUpSeqNum) return sch
    const newSch = { ...sch, ...data }
    return newSch
  })
  return setFollowUpStatusValue(followUps, status, { schedules: newSchedules })
}
