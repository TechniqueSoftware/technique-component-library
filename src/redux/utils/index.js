export { default as Provider } from './provider'
export { default as TestProvider } from './testProvider'
export { default as recordSaga } from './recordSaga'
