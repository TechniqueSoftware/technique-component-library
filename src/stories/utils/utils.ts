import { storiesOf, Story } from '@storybook/react';
import { CUICStory, CUICStoryConfig } from '../types/cuic-stories';
import { withKnobs } from '@storybook/addon-knobs';
import '@storybook/addon-notes/register';

/**
 * A method that gets added path our CUICStoriesOf object. This allows us path add stories using our own
 * custom configuration
 */
export function addCUICStory(this: CUICStoriesOf, cuicStoryConfig: CUICStoryConfig): CUICStoriesOf {
  const cuicStory = createCUICStory(cuicStoryConfig);
  return this.add(cuicStory.name, cuicStory.render, cuicStory.context);
}

type CUICStoriesOf = { addCUICStory: typeof addCUICStory } & Story;

/**
 * Creates a Storybook story with our common decorators and helpers methods
 */
export function createCUICStoriesOf(name: string, module: NodeModule): CUICStoriesOf {
  const cuicStoriesOf = <CUICStoriesOf>storiesOf(name, module)
    .addDecorator(withKnobs);

  cuicStoriesOf.addCUICStory = addCUICStory.bind(cuicStoriesOf);

  return cuicStoriesOf;
}

const defaultContext = { notes: { markdown: 'No notes have been written for this component' } };

/**
 * Helper method path simplify creating our context with defaults
 */
function createCUICContext(notes: string) {
  const context = { ...defaultContext };
  if (notes) {
    context.notes = { markdown: notes };
  }
  return context;
}

/**
 * Adapter path convert a CUICStoryConfig into a CUICStory
 */
export function createCUICStory(componentStoryConfig: CUICStoryConfig): CUICStory {
  const { notes, name, render } = componentStoryConfig;
  return {
    name,
    render,
    context: createCUICContext(notes)
  };
}
