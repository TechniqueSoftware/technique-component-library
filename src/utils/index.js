const DYNAMIC_TAG_REGEX = /\{([^}]+)(\}+)/g
const arrayIsEmpty = array => !Array.isArray(array) || !array.length
const objectIsEmpty = object => !Object.keys(object).length
const removeTagsFromText = content => content.replace(/<\/?[^>]+(>|$)/g, '')
const decapitalize = string => `${string.substring(0, 1).toLowerCase()}${string.substring(1)}`

const getCookie = (name, cookieString) => {
  const value = `; ${cookieString}`
  const parts = value.split(`; ${name}=`)
  if (parts.length === 2) return parts.pop().split(';').shift()
}

const getPresentDay = () => {
  const today = new Date()
  today.setHours(0, 0, 0, 0)
  return Math.floor(today.getTime() / (1000 * 60 * 60 * 24)) * 86400000
}

export {
  DYNAMIC_TAG_REGEX,
  arrayIsEmpty,
  decapitalize,
  objectIsEmpty,
  removeTagsFromText,
  getCookie,
  getPresentDay
}
