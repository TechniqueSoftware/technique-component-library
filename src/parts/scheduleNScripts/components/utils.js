import { STATUSES_WITH_SEGMENTS } from '@parts/scheduleNScripts/constants'
// SEGMENTS
export const checkIfAttributeIsInUse = (segments, selectedAttributes, valueToCheck, segmentId) => {
  if (valueToCheck === '') return false
  const listToCheck = segments.reduce(
    (acc, segment) => {
      if (segmentId && segment.segmentId === segmentId) return acc
      return acc.concat(segment.segmentAttributes)
    },
    selectedAttributes.map(({ value }) => ({ attributeValue: value }))
  )
  return !!listToCheck.find(({ attributeValue }) => attributeValue === valueToCheck)
}

export const getSegmentStatusProp = ({ followUpType, followUpStatus }, keyword) => {
  const segmentTp = STATUSES_WITH_SEGMENTS.find(
    s => s.type === parseInt(followUpType, 10) && s.id === parseInt(followUpStatus, 10)
  )
  return segmentTp?.[keyword]
}

export const parseAttributes = (attributes, type) => attributes.reduce((acc, attr) => {
  if (!attr.value) return acc
  return acc.concat({ attributeValue: attr.value, attributeType: type, attributeId: attr.id })
}, [])

export const unparseAttributes = attributes => attributes.map(({ attributeValue, attributeId }) => ({
  name: attributeValue,
  value: attributeValue,
  id: attributeId
}))
