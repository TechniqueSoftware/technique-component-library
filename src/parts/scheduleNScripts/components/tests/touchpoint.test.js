import React from 'react'
import { render, screen } from '@testing-library/react'
import * as redux from 'react-redux'

import userEvent from '@testing-library/user-event'

import { TestProvider as Provider } from '@redux/utils'

import TouchPoint from '../../statusSchedules/touchpoint'

const editState = {
  scheduleNScripts: {
    canEditScripts: true,
    canEditSchedules: true,
    innerPermissions: {
      automaticFollowUpAction: true,
      followUpSegmentsAction: false
    }
  }
}

const handleOptions = jest.fn()
const handleScript = jest.fn()

describe('Schedule and scripts <StatusSchedules />', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Provider>
        <TouchPoint />
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  describe('renders clock', () => {
    jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
    it('yes', () => {
      const tp = { followUpTime: '12:00' }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} />
        </Provider>
      )
      expect(screen.getByLabelText('BsClock')).toBeInTheDocument()
    })

    it('no', () => {
      render(
        <Provider>
          <TouchPoint />
        </Provider>
      )
      expect(screen.queryByLabelText('BsClock')).not.toBeInTheDocument()
    })
  })

  describe('render replay', () => {
    it('yes', () => {
      render(
        <Provider>
          <TouchPoint replay />
        </Provider>
      )
      const icon = screen.getByLabelText('BsArrowClockwise')
      expect(icon).toBeInTheDocument()
    })

    it('no', () => {
      render(
        <Provider>
          <TouchPoint />
        </Provider>
      )
      const icon = screen.queryByLabelText('BsArrowClockwise')
      expect(icon).not.toBeInTheDocument()
    })
  })

  describe('options', () => {
    it('renders', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
      render(
        <Provider>
          <TouchPoint handleOptions={handleOptions} />
        </Provider>
      )

      const dots = screen.getByLabelText('BsThreeDotsVertical')
      expect(dots).toBeInTheDocument()
      expect(dots).not.toHaveClass('hidden')
    })

    it('calls options function', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
      render(
        <Provider>
          <TouchPoint handleOptions={handleOptions} />
        </Provider>
      )
      userEvent.click(screen.queryByLabelText('BsThreeDotsVertical'))
      expect(handleOptions).toHaveBeenCalled()
    })
  })

  describe('phone scripts', () => {
    it('renders', () => {
      const tp = {
        scripts: [{ actionType: 1 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} handleScript={handleScript} />
        </Provider>
      )
      expect(screen.getByLabelText('BsTelephone')).toBeInTheDocument()
    })

    it('does not renders', () => {
      render(
        <Provider>
          <TouchPoint />
        </Provider>
      )
      expect(screen.queryByLabelText('BsTelephone')).not.toBeInTheDocument()
    })

    it('calls phone function', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
      const tp = {
        scripts: [{ actionType: 1 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} handleScript={handleScript} />
        </Provider>
      )
      userEvent.click(screen.queryByLabelText('BsTelephone'))
      expect(handleScript).toHaveBeenCalled()
    })
  })

  describe('email scripts', () => {
    it('renders', () => {
      const tp = {
        scripts: [{ actionType: 2 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} />
        </Provider>
      )
      expect(screen.getByLabelText('BsEnvelope')).toBeInTheDocument()
    })

    it('does not renders', () => {
      render(
        <Provider>
          <TouchPoint />
        </Provider>
      )
      expect(screen.queryByLabelText('BsEnvelope')).not.toBeInTheDocument()
    })

    it('calls email function', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
      const tp = {
        scripts: [{ actionType: 2 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} handleScript={handleScript} />
        </Provider>
      )
      userEvent.click(screen.queryByLabelText('BsEnvelope'))
      expect(handleScript).toHaveBeenCalled()
    })
  })

  describe('sms scripts', () => {
    it('renders', () => {
      const tp = {
        scripts: [{ actionType: 3 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} />
        </Provider>
      )
      expect(screen.getByLabelText('BsChatDots')).toBeInTheDocument()
    })

    it('does not renders', () => {
      render(
        <Provider>
          <TouchPoint />
        </Provider>
      )
      expect(screen.queryByLabelText('BsChatDots')).not.toBeInTheDocument()
    })

    it('calls sms function', () => {
      jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
      const tp = {
        scripts: [{ actionType: 3 }]
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} handleScript={handleScript} />
        </Provider>
      )
      userEvent.click(screen.queryByLabelText('BsChatDots'))
      expect(handleScript).toHaveBeenCalled()
    })
  })

  describe('no scripts', () => {
    jest.spyOn(redux, 'useSelector').mockImplementation(cb => cb(editState))
    it('renders warning signe', () => {
      render(<TouchPoint />)
      expect(screen.queryByLabelText('BsExclamationTriangleFill')).toBeInTheDocument()
    })

    it('calls handle scripts', () => {
      render(<TouchPoint handleScript={handleScript} />)
      userEvent.click(screen.queryByLabelText('BsExclamationTriangleFill'))
      expect(handleScript).toHaveBeenCalled()
    })
  })

  describe('text', () => {
    it('is immediate', () => {
      const tp = {
        followUpSeqNum: 1,
        daysTillNext: 0,
        dayNum: 0
      }
      render(<TouchPoint touchpoint={tp} />)
      expect(screen.getByText('Immediate')).toBeInTheDocument()
      expect(screen.queryByText('Day')).not.toBeInTheDocument()
    })

    it('is day', () => {
      const tp = {
        dayNum: 1,
        daysTillNext: 1
      }
      render(
        <Provider>
          <TouchPoint touchpoint={tp} />
        </Provider>
      )
      expect(screen.getByText(`Day ${tp.daysTillNext}`)).toBeInTheDocument()
      expect(screen.queryByText('Immediate')).not.toBeInTheDocument()
    })
  })
})
