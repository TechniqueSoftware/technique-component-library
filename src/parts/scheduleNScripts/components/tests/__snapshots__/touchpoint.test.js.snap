// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule and scripts <StatusSchedules /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="
    touchpoint
    filled
    rounded
    py-2
    px-3
    m-0
  "
  >
    <div
      class="row"
    >
      <p
        class="m-0 text-dark text-center fw-bold align-middle"
      >
        Day undefined
      </p>
      <div
        aria-label="BsThreeDotsVertical"
        class="
    d-flex
    justify-content-center
    align-items-center
    text-muted fs-5 
    action
  "
        role="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.5 13a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0zm0-5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"
          />
        </svg>
      </div>
    </div>
    <div
      class="gap-2 justify-content-center row"
    >
      <div>
        <div
          aria-label="BsExclamationTriangleFill"
          class="
    d-flex
    justify-content-center
    align-items-center
    text-muted fs-5
    action
  "
          role="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.982 1.566a1.13 1.13 0 0 0-1.96 0L.165 13.233c-.457.778.091 1.767.98 1.767h13.713c.889 0 1.438-.99.98-1.767L8.982 1.566zM8 5c.535 0 .954.462.9.995l-.35 3.507a.552.552 0 0 1-1.1 0L7.1 5.995A.905.905 0 0 1 8 5zm.002 6a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
